### PR TITLE
Add Py3 compatibility.

### PR DESCRIPTION
**Pull request by smurfix**
All tests pass.
Common code, no 2to3 or similar nonsense.

### CODE REVIEW

**Comment by erikrose** [comment]
Thanks for the patch! Let me have a look…

**Comment by erikrose** [comment]
First of all, adding six to the requirements in setup.py should fix the test failures.

File: parsimonious/expressions.py [hunk 1]
```diff
@@ -272,7 +273,7 @@ def _uncached_match(self, text, pos, cache, error):
     def _regex_flags_from_bits(self, bits):
         """Return the textual equivalent of numerically encoded regex flags."""
         flags = 'tilmsux'
-        return ''.join(flags[i] if (1 << i) & bits else '' for i in xrange(6))
+        return ''.join(flags[i] if (1 << i) & bits else '' for i in range(6))
```
Comment (erikrose) [inline parsimonious/expressions.py]: We should import `range` from `six.moves` so we don't materialize a list in Python 2.

File: parsimonious/grammar.py [hunk 2]
```diff
@@ -296,64 +298,74 @@ def __init__(self, custom_rules=None):
...
         Its position in the tree suffices to maintain its grouping semantics.
 
         """
+        (left_paren, _1, expression, right_paren, _2) = _a
         return expression
 
-    def visit_quantifier(self, quantifier, (symbol, _)):
+    def visit_quantifier(self, quantifier, _a):
         """Turn a quantifier into just its symbol-matching node."""
+        (symbol, _1) = _a
```
Comment (erikrose) [inline parsimonious/grammar.py]: It's such a shame that Py3 removed support for argument unpacking. I wonder if there's a more readable approach we can take.

File: parsimonious/nodes.py [hunk 3]
```diff
@@ -11,8 +11,11 @@
 
 from parsimonious.exceptions import VisitationError, UndefinedLabel
 from parsimonious.utils import StrAndRepr
+from six import reraise, python_2_unicode_compatible, with_metaclass,
+    iteritems
 
 
+@python_2_unicode_compatible
```
Comment (erikrose) [inline parsimonious/nodes.py]: Ah, where has this been all my life? :-)

File: parsimonious/expressions.py (hunk 1 above)
Comment (smurfix) [inline parsimonious/expressions.py]: On 21.09.2015 06:34, Erik Rose wrote:
>
> In parsimonious/expressions.py
> <https://github.com/erikrose/parsimonious/pull/79#discussion_r39941304>:
>
> > @@ -272,7 +273,7 @@ def _uncached_match(self, text, pos, cache, error):
> >      def _regex_flags_from_bits(self, bits):
> >          """Return the textual equivalent of numerically encoded regex flags."""
> >          flags = 'tilmsux'
> > -        return ''.join(flags[i] if (1 << i) & bits else '' for i in xrange(6))
> > +        return ''.join(flags[i] if (1 << i) & bits else '' for i in range(6))
>
> We should import |range| from |six.moves| so we don't materialize a
> list in Python 2.
>
>

Sure, but materializing a range(6) list is probably cheaper than a
generator …

-- 
-- Matthias Urlichs

Comment (smurfix) [inline parsimonious/expressions.py]: On 21.09.2015 08:16, Matthias Urlichs wrote:
> On 21.09.2015 06:34, Erik Rose wrote:
>>
>> In parsimonious/expressions.py
>> <https://github.com/erikrose/parsimonious/pull/79#discussion_r39941304>:
>>
>> > @@ -272,7 +273,7 @@ def _uncached_match(self, text, pos, cache, error):
>> >      def _regex_flags_from_bits(self, bits):
>> >          """Return the textual equivalent of numerically encoded regex flags."""
>> >          flags = 'tilmsux'
>> > -        return ''.join(flags[i] if (1 << i) & bits else '' for i in xrange(6))
>> > +        return ''.join(flags[i] if (1 << i) & bits else '' for i in range(6))
>>
>> We should import |range| from |six.moves| so we don't materialize a
>> list in Python 2.
>>
>>
>
> Sure, but materializing a range(6) list is probably cheaper than a
> generator …
… also, looking at that code, this should be len(flags), not 6 …


-- 
-- Matthias Urlichs

Comment (lucaswiman) [inline parsimonious/expressions.py]: > … also, looking at that code, this should be len(flags), not 6 …

I think it should be `range(1, len(flags))`. `t` isn't a valid regex flag (and neither is `0b1`, so it never gets included). I think it's impossible to get `x` included:
```python
>>> Regex('foo', verbose=True)._as_rhs()
'~"foo"'
>>> Regex('foo', ignore_case=True)._as_rhs()
'~"foo"i'
```
Fixed in 115539cbb590b47c2d64aae8d66209150e723709, so now the options display correctly:
```
>>> Regex('foo', ignore_case=True, locale=True,
...                  multiline=True, dot_all=True, unicode=True, verbose=True)._as_rhs()
'~"foo"ilmsux'
```

**Comment by erikrose** [comment]
Subsumed by #81

**Comment by erikrose** [comment]
And #89 is now merged! Thanks for your commits!

**Pull request closed without merging**